internal/cluster/connection.go

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cluster
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync/atomic"
+	"time"
+
+	pubcluster "github.com/hazelcast/hazelcast-go-client/cluster"
+	"github.com/hazelcast/hazelcast-go-client/internal/event"
+	ihzerror "github.com/hazelcast/hazelcast-go-client/internal/hzerror"
+	"github.com/hazelcast/hazelcast-go-client/internal/invocation"
+	ilogger "github.com/hazelcast/hazelcast-go-client/internal/logger"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto/bufutil"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto/codec"
+	"github.com/hazelcast/hazelcast-go-client/internal/util/timeutil"
+	"github.com/hazelcast/hazelcast-go-client/internal/util/versionutil"
+)
+
+const (
+	bufferSize      = 8 * 1024
+	protocolStarter = "CP2"
+)
+
+const (
+	open int32 = iota
+	closed
+)
+
+type ResponseHandler func(msg *proto.ClientMessage)
+
+type Connection struct {
+	responseCh                chan<- *proto.ClientMessage
+	pending                   chan *proto.ClientMessage
+	received                  chan *proto.ClientMessage
+	socket                    net.Conn
+	doneCh                    chan struct{}
+	endpoint                  atomic.Value
+	status                    int32
+	lastRead                  atomic.Value
+	lastWrite                 atomic.Value
+	closedTime                atomic.Value
+	writeBuffer               []byte
+	connectionID              int64
+	eventDispatcher           *event.DispatchService
+	connectedServerVersion    int32
+	connectedServerVersionStr string
+	startTime                 int64
+	logger                    ilogger.Logger
+}
+
+func (c *Connection) ConnectionID() int64 {
+	return c.connectionID
+}
+
+func (c *Connection) start(clusterCfg *pubcluster.Config, addr pubcluster.Address) error {
+	if socket, err := c.createSocket(clusterCfg, addr); err != nil {
+		return err
+	} else {
+		c.socket = socket
+		c.lastWrite.Store(time.Time{})
+		c.closedTime.Store(time.Time{})
+		c.startTime = timeutil.GetCurrentTimeInMilliSeconds()
+		c.lastRead.Store(time.Now())
+		if err := c.sendProtocolStarter(); err != nil {
+			c.socket.Close()
+			c.socket = nil
+			return err
+		}
+		go c.socketReadLoop()
+		go c.socketWriteLoop()
```

Comment:
Having a single event loop per reads and writes hurts performance since only one write on a single connection can make progress simultaneously. A better approach might be to have a number (say, `GOMAXPROCS`) goroutines responsible for reads and writes for active connections.

Another thing that could improve performance significantly is accumulation of queued client messages into a single buffer before writing them into the socket.

Reply:
I've tried having more than a single write loop, but that doesn't seem to help dramatically. Having multiple read loops doesn't seem to be feasible, unless we share the message reader between loops.

I didn't try batching, that will probably affect the performance quite dramatically.

Reply:
Sorry, I misread the code. Please ignore the above message.

nit: event loops are run on per connection basis and the problem is a bit different. When it comes to writes, the concurrency between connections is limited by the `processIncoming()` loop run by the invocation service. Since it reads from the `requestCh` channel in the loop and then writes the client message into a channel of size 1 on one of connections, a "slow" connection or a write of a large payload on one of connections may lead to a situation when `requestCh` is full and no operations are able to make progress. A straight-forward solution to this problem may be to allow caller goroutines to write into connections directly, rather than through a set of event loops connected with channels.